Create how-to-detect-win11.md (#1505)
MicrosoftDocs/edge-developer: microsoft-edge/web-platform/how-to-detect-win11.md

---
description: How to differentiate Windows 10 and Windows 11 using User-Agent Client Hints
title: Detecting Windows 11 using User-Agent Client Hints
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 10/15/2021
ms.topic: article
ms.prod: microsoft-edge
keywords: microsoft edge, compatibility, web platform, user-agent string, ua string, user-agent client hints, user agent client hints, ua client hints, ua ch, windows 11, win11, detecting windows 11, windows detection
---
# Detecting Windows 11 using User-Agent Client Hints

<!--
Restrict the lexicon to these forms:
User-Agent string
user agent string

> User-Agent Client Hints

Reviewer: ```suggestion
User-Agent Client Hints
user agent
```

| Header field | Values that indicate Windows 10 | Values that indicate Windows 11 |
| --- | --- | --- |
| `Sec-CH-UA-Platform` | `Windows` | `Windows` |
| `Sec-CH-UA-Platform-Version` | values between `1.0.0` and `10.0.0` | `13.0.0` and above |

User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10.  We don't recommend using User-Agent strings to retrieve user agent data.  Browsers that do not support User-Agent Client Hints will not be able to differentiate between Windows 11 and Windows 10.

<!-- ====================================================================== -->
## Browsers that support User-Agent Client Hints

The following table shows which browsers support differentiating between Windows 11 and Windows 10.

| Browser | Supports differentiation via User-Agent Client Hints? | Alternative detection methods |
| --- | --- | --- |
| Microsoft Edge 94+ | Yes | - |
| Chrome 95+ | Yes | - |
| Opera | Yes | - |
| Firefox | No | None |
| Internet Explorer 11 | No | None |

> | Microsoft Edge Legacy | No | window.external.getHostEnvironmentValue |

Reviewer 2: Let's remove this line entirely. Support for Edge Legacy ended on March 9, 2021. We should not document its support/non-support for things at this point. I briefly thought there might be some applicability for WWA apps using the Edge Legacy platform, but the getHostEnvironmentValue API was never available to those.

> ---

Reviewer 3: Can we add an entry to the ToC file because this is a new file?

description: How to differentiate Windows 10 and Windows 11 using User-Agent Client Hints
title: Detecting Windows 11 using User-Agent Client Hints
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 10/15/2021
ms.topic: article
ms.prod: microsoft-edge
keywords: microsoft edge, compatibility, web platform, user-agent string, ua string, user-agent client hints, user agent client hints, ua client hints, ua ch, windows 11, win11, detecting windows 11, windows detection
---
# Detecting Windows 11 using User-Agent Client Hints

<!--
Restrict the lexicon to these forms:
User-Agent string
user agent string
User-Agent Client Hints
-->

Websites can use user agent information sent from the browser to detect brand, version, device platform, and more. There are two approaches for sites to access user agent information:
*  User-Agent strings (legacy).
*  User-Agent Client Hints (recommended).

For details about these two approaches, see [Detecting Microsoft Edge from your website](user-agent-guidance.md).

> To differentiate between users who are using Windows 11 and Windows 10, starting in Microsoft Edge version 94 and Chrome version 95, User-Agent Client Hints reflects the following header-field values:

Reviewer 4: This phrasing is a bit awkward. Consider:
```suggestion
Starting in Microsoft Edge version 94 and Chrome version 95, sites can differentiate between users on Windows 11 and Windows 10 via User-Agent Client Hints. This information can be found in the following UA-CH request headers:
```

| Header field | Values that indicate Windows 10 | Value that indicates Windows 11 |
| --- | --- | --- |
| `Sec-CH-UA-Platform` | `Windows` | `Windows` |
| `Sec-CH-UA-Platform-Version` | values between `1.0.0` and `10.0.0` | `13.0.0` and above |

> User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10.  We don't recommend using User-Agent strings to retrieve user agent data.  Browsers that do not support User-Agent Client Hints will not be able to differentiate between Windows 10 and Windows 11.

Reviewer 4: nit: We say "Windows 11 and Windows 10" elsewhere. Maybe we should do that here too:
```suggestion
User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10.  We don't recommend using User-Agent strings to retrieve user agent data.  Browsers that do not support User-Agent Client Hints will not be able to differentiate between Windows 11 and Windows 10.
```

<!-- ====================================================================== -->
## Browsers that support User-Agent Client Hints

The following table shows which browsers support differentiating between Windows 11 and Windows 10.

| Browser | Supports differentiation via User-Agent Client Hints? | Alternative detection methods |
| --- | --- | --- |
| Microsoft Edge 94+ | Yes | - |
| Chrome 95+ | Yes | - |
| Opera | Yes | - |
| Firefox | No | None |
| Internet Explorer 11 | No | None |
| Microsoft Edge Legacy | No | window.external.getHostEnvironmentValue |
| Safari | No | Not Applicable |

<!-- ====================================================================== -->
## Sample code for detecting Windows 11

The following code detects Windows 11:

```javascript
navigator.userAgentData.getHighEntropyValues(
 ["platform",
  "platformVersion"])
 .then(ua => {
   if (ua.platform === "Windows") {
     const majorPlatformVersion = ua.platformVersion.split('.')[0];
     if (majorPlatformVersion >= 13) {
       console.log("Windows 11+");
      }
      else if (majorPlatformVersion > 0) {
        console.log("Windows 10");
      }
      else {

>         console.log("Pre-Windows 10");

Reviewer 3: ```suggestion
        console.log("Before Windows 10");
```

      }
   }
   else {
     console.log("Not running on Windows");
   }
 });

```

<!-- ====================================================================== -->
## Detecting previous Windows versions

> To detect older versions of Windows, use the following values for `platformVersion` in User-Agent Client Hints:

Reviewer 4: This is a table of a bunch of versions of Windows, including Window 11. Maybe something like this makes sense:
```suggestion
To detect specific versions of Windows, use the following values for `platformVersion` in User-Agent Client Hints:
```

---
description: How to detect Microsoft Edge data with User-Agent Client Hints and User-Agent strings.
title: Detecting Windows 11 using User-Agent Client Hints
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 10/15/2021
ms.topic: article
ms.prod: microsoft-edge
keywords: microsoft edge, compatibility, web platform, user-agent string, ua string, user-agent client hints, user agent client hints, ua client hints, ua ch, windows 11, win11, detecting windows 11, windows detection
---
# Detecting Windows 11 using User-Agent Client Hints

<!--
Restrict the lexicon to these forms:
User-Agent string
user agent string
User-Agent Client Hints
-->

> Websites can use User Agent information that's sent from the browser to detect information such as browser brand, version, and platform. There are two approaches for sites to access User Agent information:

Reviewer 4: > information sent from the browser to detect information such as browser brand

nit: Redundant use of "information". Maybe just:

> information sent from the browser to detect brand

> Websites can use User Agent information that's sent from the browser to detect information such as browser brand, version, and platform. There are two approaches for sites to access User Agent information:

Reviewer 2: When not specifically referring to the User-Agent Client Hints proposal, we should not capitalize "user agent" as it's not a proper name. i.e. in these two sentences it should not be capitalized.

*  User-Agent strings (legacy).
*  User-Agent Client Hints (recommended).

For details about these two approaches, see [Detecting Microsoft Edge from your website](user-agent-guidance.md).

To differentiate between users who are using Windows 11 and Windows 10, starting in Microsoft Edge version 94, User-Agent Client Hints reflects the following header-field values:

| Header field | Value that indicates Windows 10 | Value that indicates Windows 11 |
| --- | --- | --- |
| `Sec-CH-UA-Platform` | `Windows` | `Windows` |

> | `Sec-CH-UA-PlatformVersion` | `10.0.0` and above | `13.0.0` and above |

Reviewer 2: Sec-CH-UA-PlatformVersion -> Sec-CH-UA-Platform-Version

> Unfortunately, User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10.  We don't recommend using User-Agent strings to retrieve user agent data.  For browsers which don't support User-Agent Client Hints, sites won't be able to differentiate between Windows 11 and Windows 10.

Reviewer 4: > For browsers who do not support UA Client Hints, sites won't be able to differentiate between Windows 10 and Windows 11

Suggested rephrase:

> Browsers that do not support User Agent Client Hints will not be able to differentiate between Windows 10 and Windows 11.

> Unfortunately, User-Agent strings won't be updated to differentiate between Windows 11 and Windows 10.  We don't recommend using User-Agent strings to retrieve user agent data.  For browsers which don't support User-Agent Client Hints, sites won't be able to differentiate between Windows 11 and Windows 10.

Reviewer 4: I don't think we should say "Unfortunately" here

<!-- ====================================================================== -->
## Browsers that support User-Agent Client Hints

> The following table shows which browsers support differentiating between Windows 11 and Windows 10.

Reviewer: @jenny-microsoft - Can you use fewer than 3 different values?  Currently the column has several values; unclear:
`-`
`None`
`N/A`

Reviewer: @jenny-microsoft - Yes, I think inserting that text directly would be more helpful, since there are 3 similar looking notations now.

Reviewer 2: We could also just remove Safari from the list. It's interesting to list it in the context of broader UA Client Hints support, but for a page dedicated to differentiating Win10 vs. Win11 it's unncessary.

Reviewer: Resolution: Removed the now-empty "Alternative" column.

| Browser | Supports User-Agent Client Hints? | Alternative detection methods |
| --- | --- | --- |
| Microsoft Edge 94+ | Yes | - |

> | Chrome 94+ | Yes | - |

Reviewer 2: Chrome 94+ doesn't make sense here-- it's supported UA Client Hints for a long time, but not the new definition of platform version that allows differentiating Win10 and Win11.

Perhaps we should update this to "Chrome 95+" and update the column heading to say "Supports latest definition for UA Client Hints platform version?"

| Opera | Yes | - |
| Firefox | No | None |
| Internet Explorer 11 | No | None |
| Microsoft Edge Legacy | No | window.external.getHostEnvironmentValue |
| Safari | No | N/A |

<!-- ====================================================================== -->
## Sample code for detecting Windows 11

The following code detects Windows 11:

```javascript
navigator.userAgentData.getHighEntropyValues(

>  ["platform",

Reviewer 2: Platform is a low-entropy hint and doesn't need to be explicitly specified-- it will be available even if not present.

Reviewer 2: Ah, I was looking at the spec and thought it would. Yes, let's stick with looking at navigator.userAgentData.platform for the Windows check.

  "platformVersion"])
 .then(ua => {
   if (ua.platform === "Windows" && ua.platformVersion.split('.')[0] >= 13) {
     console.log("Windows 11+");
   }
   else if (ua.platform === "Windows") {
     console.log("Windows 10 or lower");
   }
 });

```

> [!NOTE]

> > We recommend that you don't use `parseFloat()` to perform `platformVersion` checks, because potential formatting changes could lead to incorrect results.

Reviewer 4: This note seems unnecessary since we only have a "good" example above

Reviewer 2: Right, we don't need to document all of the potential ways you could author it poorly.
